### ISSUE TITLE
Implement activation params type

### ISSUE THREAD

**Issue by milancurcic** [modern-fortran/neural-fortran#124]
Some activation functions like leaky ReLU (#123) require one or more additional parameters.

To allow passing activation functions as procedure pointers, all functions must have the same interface. A proposed general solution (thank @jvdp1) is to:

1. Define a derived type `activation_params` or similar that defines any possible extra parameters that may be needed by activation functions; set default values of activation parameters in the type definition.
  ```fortran
type :: activation_params
  real :: alpha = 0.3
end type activation_params
```
2. Make each activation function have a `type(activation_params), intent(in) optional :: params` dummy argument (instead of the current `alpha`). Inside activation function definitions, function that use one or more activation parameters access them directly; those that don't simply ignore it.
```fortran
  pure function leaky_relu(x, params) result(res)
    !! Leaky Rectified Linear Unit (Leaky ReLU) activation function.
    real, intent(in) :: x(:)
    type(activation_params), intent(in), optional :: params
    real :: res(size(x))
    res = max(params % alpha * x, x)
  end function leaky_relu
```
3. Make `activation_params` an attribute of `dense` and `conv2d` layers (and later any other layers that activate).
4. Pass the attributes to the activation procedure associated with the layer.

**Comment by milancurcic** [comment]
Thanks @ggoyman, I believe [inference-engine](https://github.com/BerkeleyLab/inference-engine) takes a similar approach.

In a nutshell, it seems to me that an abstract class approach allows the activation-specific parameters to be carried with the concrete activation type itself, rather than the layer type. I like that.

Would you be open to contributing this as a PR? I'd help.

**Comment by rouson** [comment]
Yes,that's the Inference-Engine approach.  We call the abstract type activation_strategy_t because it's an example of the [Strategy design pattern ](https://en.wikipedia.org/wiki/Strategy_pattern?wprov=sfti1).

**Comment by ggoyman** [comment]
@milancurcic, OK, I'll try to implement this solution.

**Comment by milancurcic** [comment]
Solved by #126.
